ext/standard/array.c

```diff
@@ -6468,11 +6470,16 @@ PHP_FUNCTION(array_reduce)
 	}
 
 	fci.retval = &retval;
-	fci.param_count = 2;
+	fci.param_count = 3;
 
-	ZEND_HASH_FOREACH_VAL(htbl, operand) {
+	ZEND_HASH_FOREACH_KEY_VAL(htbl, num_key, string_key, operand) {
 		ZVAL_COPY_VALUE(&args[0], return_value);
 		ZVAL_COPY(&args[1], operand);
+		if (!string_key) {
+			ZVAL_LONG(&args[2], num_key);
+		} else {
+			ZVAL_STR_COPY(&args[2], string_key);
```

Comment:
Because you now make a string copy into `args[2]`, you need to call `zval_ptr_dtor(&args[2])` at the right places, otherwise you risk leaking memory. The reason you're not observing the memory leak now in your test is because your test uses interned strings which can't leak per definition.
To create a non-interned string, I use the trick: `str_repeat("my string here", random_int(1,1))` which will defeat any optimization regarding interned strings.

Reply:
Oh, I saw those calls in `array_filter`, but I didn't know what they were for, and I guess I still don't know how to use them...

Reply:
OK, I think I figured it out. I'll push an update later when I've done some more testing.

Reply:
> Is it the case that the test framework will detect such memory leaks (with non-interned strings) itself?

It's not the test framework that will detect it, but debug builds of PHP themselves.
Debug builds of PHP check at the end of a request whether all memory was freed and print a warning if it's not the case. This only works though if the environment variable `USE_ZEND_ALLOC` is unset or set to 1.
Furthermore, CI runs the tests under ASAN which gives a very neat report of potential memory leaks.